.Lib9c.Tests/Action/HackAndSlashTest.cs

```diff
@@ -1018,6 +1016,146 @@ public void CheckRewardItems(bool backward, int worldId, int stageId)
+
+            state = state.SetState(
+                _avatarAddress.Derive("world_ids"),
+                List.Empty.Add(worldId.Serialize())
+            );
+
+            var skillStateAddress = Addresses.GetSkillStateAddressFromAvatarAddress(_avatarAddress);
+            CrystalRandomSkillState skillState = null;
+            if (hasCrystalSkill)
+            {
+                skillState = new CrystalRandomSkillState(skillStateAddress, stageId);
+                skillState.Update(int.MaxValue, _tableSheets.CrystalStageBuffGachaSheet);
+                state = state.SetState(skillStateAddress, skillState.Serialize());
+            }
+
+            var action = new HackAndSlash
+            {
+                costumes = clear ? costumes : new List<Guid>(),
+                equipments = clear
+                    ? equipments.Select(e => e.NonFungibleId).ToList()
+                    : new List<Guid>(),
+                foods = new List<Guid>(),
+                worldId = worldId,
+                stageId = stageId,
+                avatarAddress = _avatarAddress,
+                stageBuffId = skillState?.SkillIds
+                    .OrderBy(key => _tableSheets.CrystalRandomBuffSheet[key].Rank)
+                    .FirstOrDefault(),
+            };
+
+            var ctx = new ActionContext
+            {
+                PreviousStates = state,
+                Signer = _agentAddress,
+                Random = new TestRandom(),
+                Rehearsal = false,
+                BlockIndex = 1,
+            };
+            var nextState = action.Execute(ctx);
+            var simulator = new StageSimulator(
+                new TestRandom(ctx.Random.Seed),
+                previousAvatarState,
+                new List<Guid>(),
+                worldId,
+                stageId,
+                _tableSheets.GetStageSimulatorSheets(),
+                _tableSheets.CostumeStatSheet,
+                StageSimulator.ConstructorVersionV100080);
+            simulator.Simulate(1);
+            var log = simulator.Log;
+            var skillStateIValue =
+                nextState.GetState(skillStateAddress);
+            var serialized = skillStateIValue as List;
+            Assert.NotNull(serialized);
+            skillState = new CrystalRandomSkillState(skillStateAddress, serialized);
+            Assert.Equal(skillStateAddress, skillState.Address);
+
+            if (clear)
+            {
+                Assert.Equal(stageId + 1, skillState.StageId);
+                Assert.Equal(0, skillState.StarCount);
+            }
+            else
+            {
+                Assert.Equal(stageId, skillState.StageId);
+                Assert.Equal(
+                    hasCrystalSkill
+                        ? _tableSheets.CrystalStageBuffGachaSheet[stageId].MaxStar
+                        : log.clearedWaveNumber,
+                    skillState.StarCount);
+            }
+
+            Assert.Empty(skillState.SkillIds);
```

Comment:
it should be check also clear case.

Reply:
It already check clear case and failed case.